third_party/terraform/resources/resource_compute_router_nat.go

```diff
@@ -78,80 +28,109 @@ func resourceComputeRouterNat() *schema.Resource {
 		},
 
 		Schema: map[string]*schema.Schema{
-			"name": &schema.Schema{
+			"name": {
 				Type:         schema.TypeString,
 				Required:     true,
 				ForceNew:     true,
 				ValidateFunc: validateRFC1035Name(2, 63),
 			},
-			"router": &schema.Schema{
+			"router": {
 				Type:     schema.TypeString,
 				Required: true,
 				ForceNew: true,
 			},
-			"nat_ip_allocate_option": &schema.Schema{
+			"nat_ip_allocate_option": {
 				Type:         schema.TypeString,
 				Required:     true,
-				ForceNew:     true,
 				ValidateFunc: validation.StringInSlice([]string{"MANUAL_ONLY", "AUTO_ONLY"}, false),
 			},
-			"nat_ips": &schema.Schema{
+			"nat_ips": {
 				Type:     schema.TypeSet,
 				Optional: true,
-				ForceNew: true,
 				Elem:     &schema.Schema{Type: schema.TypeString},
 			},
-			"source_subnetwork_ip_ranges_to_nat": &schema.Schema{
+			"source_subnetwork_ip_ranges_to_nat": {
 				Type:         schema.TypeString,
-				Optional:     true,
-				ForceNew:     true,
+				Required:     true,
 				ValidateFunc: validation.StringInSlice([]string{"ALL_SUBNETWORKS_ALL_IP_RANGES", "ALL_SUBNETWORKS_ALL_PRIMARY_IP_RANGES", "LIST_OF_SUBNETWORKS"}, false),
 			},
-			"subnetwork": &schema.Schema{
+			"subnetwork": {
 				Type:     schema.TypeSet,
 				Optional: true,
-				ForceNew: true,
-				Elem:     routerNatSubnetworkConfig,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+
+						// this field is optional with a default in the API, but we
+						// don't have the ability to support complex defaults inside
+						// nested fields
+						"source_ip_ranges_to_nat": {
+							Type:     schema.TypeSet,
+							Required: true,
+							MinItems: 1,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+						"secondary_ip_range_names": {
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+						},
+					},
+				},
 			},
-			"min_ports_per_vm": &schema.Schema{
+			"min_ports_per_vm": {
 				Type:     schema.TypeInt,
 				Optional: true,
-				ForceNew: true,
+				Computed: true,
 			},
-			"udp_idle_timeout_sec": &schema.Schema{
+			"udp_idle_timeout_sec": {
```

Comment:
I think this field and the other O+C ones below have well-defined defaults. I'm assuming the API returns `nil` if they take the default; does it do the same when explicit values are sent?

If we could `default_if_empty` these fields and set Terraform-side defaults, I'd feel a lot better than declaring more O+C fields.

Reply:
the API JSON response doesn't have the fields but golang library means it returns zeros - your linked issue though suggested we shouldn't because its a breaking change to set defaults? so we might need to do it in v3

I'm also fine leaving it as just computed for now and making changes in 3.0.0 since I don't think it changes the behavior that much since rn the API doesn't return anything

Reply:
Oh no that totally means explicit `0` and the default are indistinguishable doesn't it. Sounds good to me.